File: KwizzadExample/KwizzadExample/DebugViewController.m
//
//  DebugViewController.m
//  KwizzadExample
//
//  Created by Fares Ben Hamouda on 27.04.17.
//  Copyright © 2017 Kwizzad. All rights reserved.
//

#import "DebugViewController.h"

@interface DebugViewController ()

@end

@implementation DebugViewController

KwizzadSDK* kwizzadInstance;

- (void)viewDidLoad {
    [super viewDidLoad];
    [self clearLog];
}

-(void) viewDidAppear:(BOOL)animated {
    kwizzadInstance = KwizzadSDK.instance ;
    kwizzadInstance.delegate = self;
    [self preloadButtonPressed:self];
}

- (IBAction)preloadButtonPressed:(id)sender {
    // Fade adview while preloading ad
    _preloadButton.enabled = NO;
    _adView.alpha = 0.3;
    _adView.headlineLabel.text = @"Loading..." ;
    _adView.detailsLabel.text = @"Loading..." ;
    
    if ( [_placementField.text  isEqual: @""]) {
        return;
    } else {
        [kwizzadInstance requestAdWithPlacementId:_placementField.text onAdAvailable:nil];
        _kwizzadViewController = nil;
        [self clearLog];
        [_placementField resignFirstResponder];
    }
}

- (IBAction)startAd:(id)sender {
    if ( [kwizzadInstance canShowAdWithPlacementId:_placementField.text]) {
        [self presentViewController:_kwizzadViewController animated:YES completion:nil];
    }
}

-(void) loadImage:(NSString*)urlString {
    NSURL *url = [NSURL URLWithString:urlString];
    NSData *data = [NSData dataWithContentsOfURL:url];
    UIImage *image = [UIImage imageWithData:data];
    [_adView.imageView setImage:image];
}

/////////////////////////////////////////////////////////////////////
//                                                                 //
//  The following delegate methods are called by the Kwizzad SDK.  //
//                                                                 //
/////////////////////////////////////////////////////////////////////

- (void) kwizzadDidRequestAdWithPlacementId:(NSString *)placementId {
    [self log:[NSString stringWithFormat:@"Requested an ad on placement %@",placementId]];
}

-(void) kwizzadOnAdAvailableWithPlacementId:(NSString *)placementId potentialRewards:(NSArray<KwizzadReward *> *)potentialRewards adResponse:(KwizzadAdResponseEvent *)adResponse {
    [self log:[NSString stringWithFormat:@"A new ad is available on placement %@",placementId]];
    [self.adView setHidden:NO];
    
    // For better targeting, you can set known user data here:
    KwizzadUserDataModel *userData = kwizzadInstance.userDataModel;
    userData.userName = @"Francesca Rossi"; // user name inside your app
    userData.facebookUserId = @"1234abc";
    userData.userId = @"12345"; // identifies the user inside your app
    userData.gender = KwizzadGenderFemale;
    
    NSDictionary<NSString * ,id> *customParameters = [[NSDictionary alloc]init] ;// Optionally, you can supply custom parameters for special use cases here.
    
    _kwizzadViewController = [kwizzadInstance loadAdWithPlacementId:placementId customParameters:customParameters];
    
    int thumbnailWidth = 400;
    NSString *thumbnailUrlString = [adResponse squaredThumbnailURLWithWidth:thumbnailWidth];

    [self loadImage:thumbnailUrlString];
    
    // If you want, you can also make use of the campaign's teaser text, headline and brand name.
    // All of these values are optional, so ensure your code works when they are `nil`.
    KwizzadAdMetaInfo *adMetaInfo = adResponse.adMetaInfo;
    [self log:[NSString stringWithFormat:@"Teaser text: %@", adMetaInfo.teaser]];
    [self log:[NSString stringWithFormat:@"Brand name: %@", adMetaInfo.brand]];
    [self log:[NSString stringWithFormat:@"Headline text: %@", adMetaInfo.headline]];

    _adView.headlineLabel.text = adMetaInfo.headline ;
    _adView.detailsLabel.text = adMetaInfo.teaser;
    _adView.smilesLabel.text = [NSString stringWithFormat:@"+%@",[[KwizzadReward summarizeWithRewards:potentialRewards].firstObject valueOrMaxValue]];

    [_adView layoutSubviews];
    self.height.constant = [_adView preferredHeight];
    [_adView layoutSubviews];
    
    // Optionally, you can use this to show the user their potential reward(s).
    // TODO: Generate rewards string correctly, port from Web SDK
    NSString *rewardStrings = @"";
    for ( KwizzadReward *reward in potentialRewards) {
        NSString *rewardString = [reward valueDescription];
        rewardStrings = [rewardStrings stringByAppendingString:[NSString stringWithFormat:@"%@ ,",rewardString]];
    }
    [self log:[NSString stringWithFormat:@"Potential rewards: %@",rewardStrings]];
}

- (void) kwizzadOnAdReadyWithPlacementId:(NSString *)placementId {
    [self log:[NSString stringWithFormat:@"The ad on placement %@ is loaded / ready to be displayed.",placementId]];
    _preloadButton.enabled = YES;
    _adView.alpha = 1.0;
}

- (void) kwizzadDidShowAdWithPlacementId:(NSString *)placementId {
    [self log:[NSString stringWithFormat:@"The ad on placement %@ is shown.",placementId]];
}

- (void) kwizzadDidDismissAdWithPlacementId:(NSString *)placementId {
    [self log:[NSString stringWithFormat:@"The ad on placement %@ has been dismissed.",placementId]];
}

- (void) kwizzadGotOpenTransactionsWithOpenTransactions:(NSSet<KwizzadOpenTransaction *> *)openTransactions {
    [self log:[NSString stringWithFormat:@"%lu incoming transactions to be confirmed.",(unsigned long)openTransactions.count]];
    for (KwizzadOpenTransaction * t in openTransactions) {
        [self log:t.description];
        if ( t.reward) {
            NSString *msg = @"You've earned a reward!";
            if ( t.reward.amount && t.reward.currency) {
                msg = [NSString stringWithFormat:@"You've earned %@ %@",t.reward.amount ,t.reward.currency];
            }
            
            UIAlertController *alert = [UIAlertController alertControllerWithTitle: @"Congratulations!"
                                                                           message: msg
                                                                    preferredStyle: UIAlertControllerStyleAlert];
            
            [alert addAction: [UIAlertAction actionWithTitle:@"Yo" style: UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                [kwizzadInstance completeWithTransaction:t];
            }]];
            
            [self presentViewController:alert animated:YES completion:nil];
            
        }
        else {
            // Silently confirm the transaction
            [kwizzadInstance completeWithTransaction:t];
        }
    }
}

// optional
-(void) kwizzadOnNoFillWithPlacementId:(NSString *)placementId {
    [self log:[NSString stringWithFormat:@"Received a no-fill response on placement %@",placementId]];
    [self.adView setHidden: YES];
}

// optional
-(void) kwizzadOnErrorOccuredWithReason:(NSString *)reason placementId:(NSString *)placementId {
    [self log:[NSString stringWithFormat:@"An error occured on placement %@: %@",placementId,reason]];
}

// optional
-(void) kwizzadWillPresentAdWithPlacementId:(NSString *)placementId {
    [self log:[NSString stringWithFormat:@"Kwizzad will present an ad on placement %@",placementId]];
}

// optional
-(void) kwizzadWillDismissAdWithPlacementId:(NSString *)placementId {
    [self log:[NSString stringWithFormat:@"Kwizzad will dismiss an ad on placement %@",placementId]];
}

// optional
-(void) kwizzadOnGoalReachedWithPlacementId:(NSString *)placementId {
    [self log:[NSString stringWithFormat:@"User reached the campaign goal on placement %@",placementId]];
}

// optional
-(void) kwizzadCallToActionClickedWithPlacementId:(NSString *)placementId {
    [self log:[NSString stringWithFormat:@"User reached the call-to-action on placement %@",placementId]];
}

/////////////////////////////////////////////////////////////////////

// Hide keyboard when click return.
-(BOOL)textFieldShouldReturn:(UITextField *)textField {
    [self preloadButtonPressed:self];
    [textField resignFirstResponder];
    return true;
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [_placementField resignFirstResponder];
}

-(void) clearLog {
    _debugMessage.text = @"";
}

-(void) log:(NSString*)message {
    _debugMessage.text = [_debugMessage.text stringByAppendingString:[NSString stringWithFormat:@"\n\n%@",message]];
    CGPoint bottomOffset = CGPointMake(0, self.debugMessage.contentSize.height - self.debugMessage.bounds.size.height);
    [self.debugMessage setContentOffset:bottomOffset animated:YES];
    NSLog(@"%@", message);
}

@end

File: KwizzadExample/KwizzadExample/DebugViewController.swift
//
//  DebugViewController.swift
//  KwizzadExample
//
//  Created by Fares Ben Hamouda on 27.04.17.
//  Copyright © 2017 Kwizzad. All rights reserved.
//

import UIKit
import KwizzadSDK

class DebugViewController: UIViewController, KwizzadSDKDelegate, UITextFieldDelegate {
    let kwizzad = KwizzadSDK.instance;
    var kwizzadController : UIViewController?
    
    @IBOutlet var adView: AdView!
    @IBOutlet var debugMessage: UITextView!
    @IBOutlet var placementField: UITextField!
    @IBOutlet var height: NSLayoutConstraint!
    @IBOutlet var preloadButton: UIButton!

    override func viewDidLoad() {
        self.clearLog()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        kwizzad.delegate = self;
        self.preloadButtonPressed(self)
    }
    
    func loadImage(urlString: String?) {
        if let urlString = urlString {
            if let url = URL(string: urlString) {
                DispatchQueue.global().async {
                    if let data = try? Data(contentsOf: url) {
                        DispatchQueue.main.async {
                            let image = UIImage(data: data)
                            self.adView.imageView?.image = image
                        }
                    }
                }
            }
        }
    }

    @IBAction func startAd(_ sender: Any) {
        if (kwizzad.canShowAd(placementId:self.placementField.text!)) {
            guard let controller = self.kwizzadController else {
                return
            }
            self.present(controller, animated: true, completion: {
                self.log("Kwizzad controller presented.")
            })
        }
    }
    
    @IBAction func preloadButtonPressed(_ sender: Any) {
        self.clearLog()
        // Fade adview while preloading ad
        self.adView.alpha = 0.3
        self.adView.detailsLabel.text = "Loading..."
        self.adView.headlineLabel.text = "Loading..."
        preloadButton.isEnabled = false
        
        guard placementField.text != "" else {
            return
        }
        kwizzad.requestAd(placementId: placementField.text!)
        self.kwizzadController = nil
        view.endEditing(true)
    }

    func stopAd() {
        // Optionally, you can use this method to close the ad programmatically.
        KwizzadSDK.instance.close(placementField.text!)
    }

    /////////////////////////////////////////////////////////////////////
    //                                                                 //
    //  The following delegate methods are called by the Kwizzad SDK.  //
    //                                                                 //
    /////////////////////////////////////////////////////////////////////
    
    func kwizzadDidRequestAd(placementId: String) {
        self.log("Requested an ad on placement \(placementId).")
    }
    
    func kwizzadOnAdAvailable(placementId: String, potentialRewards: [Reward], adResponse: AdResponseEvent) {
        self.log("A new ad is available on placement \(placementId).")
        self.adView.isHidden = false

        // For better targeting, you can set known user data here:
        let userData = kwizzad.userDataModel;
        userData.userId = "12345" // identifies the user inside your app
        userData.gender = Gender.Female
        userData.userName = "Francesca Rossi" // user name inside your app
        userData.facebookUserId = "1234abc"

        let myCustomParameters : [String:Any] = [:]; // Optionally, you can supply custom parameters for special use cases here.
        self.kwizzadController = self.kwizzad.loadAd(placementId: placementId, customParameters: myCustomParameters)
        
        // Load the campaign image thumbnail in given given size, if possible.
        let thumbnailWidth = 400;
        let thumbnailUrlString = adResponse.squaredThumbnailURL(width: thumbnailWidth);
        self.loadImage(urlString: thumbnailUrlString)

        // If you want, you can also make use of the campaign's teaser text, headline and brand name.
        // All of these values are optional, so ensure your code works when they are `nil`.
        let adMetaInfo = adResponse.adMetaInfo;
        self.log("Brand name: \(adMetaInfo.brand ?? "(none)")");
        self.log("Teaser for campaign's content: \(adMetaInfo.teaser ?? "(none)")");

        let incentiveText = Reward.incentiveTextFor(rewards: potentialRewards)
        self.log("Incentive text: \(incentiveText ?? "(undefined)")")

        // Optionally, can use this to build your own strings that describe the potential reward(s).
        let rewardSummary = Reward.enumerateRewardsAsText(rewards: potentialRewards)
        self.log("Potential rewards: \(rewardSummary)")
        
        adView.headlineLabel.text = adMetaInfo.headline ?? "Play a quiz!";
        adView.detailsLabel.text = incentiveText ?? "";
        adView.smilesLabel.text = "+\(String(describing: Reward.summarize(rewards: potentialRewards).first!.valueOrMaxValue()!.stringValue))"
        adView.layoutSubviews()
        self.height.constant = adView.preferredHeight()
        adView.layoutSubviews()
        
        self.log("Ad will expire in \(String(describing: adResponse.expiry?.timeIntervalSinceNow)) seconds.");
    }

    func kwizzadOnAdReady(placementId: String) {
        self.log("The ad on placement \(placementId) is loaded / ready to be displayed.")
        self.adView.alpha = 1
        self.preloadButton.isEnabled = true
    }
    
    func kwizzadDidShowAd(placementId: String) {
        self.log("The ad on placement \(placementId) is shown.")
    }
    
    func kwizzadDidDismissAd(placementId: String) {
        self.log("The ad on placement \(placementId) has been dismissed.")
    }
    
    func kwizzadGotOpenTransactions(openTransactions: Set<OpenTransaction>) {
        self.log("\(openTransactions.count) incoming transactions to be confirmed.")
        for transaction in openTransactions {
            self.log(transaction.description)
            if let reward = transaction.reward {
                var msg = "You've earned a reward!"
                if let amount = reward.amount, let currency = reward.currency {
                    msg = "You've earned \(amount) \(currency)."
                }
                let alert = UIAlertController(title: "Congratulations!", message: msg, preferredStyle: UIAlertControllerStyle.alert)

                alert.addAction(UIAlertAction(title: "Yo", style: UIAlertActionStyle.default, handler: { foo in
                    self.kwizzad.complete(transaction: transaction)
                }))

                self.present(alert, animated: true, completion: nil)
            }
            else {
                self.kwizzad.complete(transaction: transaction) // Silently confirm the transaction
            }
        }
    }
    
    func kwizzadOnNoFill(placementId: String) {
        self.adView.isHidden = true
        self.preloadButton.isEnabled = true
        self.log("Received a no-fill response on placement \(placementId).");
    }
    
    // optional
    func kwizzadOnErrorOccured(placementId: String, reason: String) {
        self.log("An error occured on placement \(placementId). Reason : \(reason))");
        self.adView.isHidden = true
        self.preloadButton.isEnabled = true
    }
    
    // optional
    func kwizzadWillPresentAd(placementId: String) {
        self.log("Kwizzad will present an ad on placement \(placementId).");
    }
    
    // optional
    func kwizzadWillDismissAd(placementId: String) {
        self.log("Kwizzad will dismiss an ad on placement \(placementId).");
    }
    
    // optional
    func kwizzadOnGoalReached(placementId: String) {
        self.log("User reached the campaign goal on placement \(placementId).");
    }
    
    // optional
    func kwizzadCallToActionClicked(placementId: String) {
        self.log("User reached the call-to-action on placement \(placementId).");
    }
    
    /////////////////////////////////////////////////////////////////////
    
    // Hide keyboard when click return.
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        preloadButtonPressed(self)
        textField.resignFirstResponder()
        return true;
    }
    
    // Hide keyboard when click outside.
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        placementField.resignFirstResponder()
        self.view.endEditing(true)
    }
    
    func log(_ message: String) {
        let view = self.debugMessage!
        view.text.append("\n\n\(message)")
        let bottom = view.contentSize.height - view.bounds.size.height
        view.setContentOffset(CGPoint(x: 0, y: bottom), animated: true)
        print(message);
    }
    
    func clearLog() {
        self.debugMessage.text = "";
    }
}